### PR TITLE
BarrierEsp

### PR DESCRIPTION
**Pull request by C0derK1d**
## Description
#715 

**Warning:** It's only works when the **Particles** is set to **all**

### CODE REVIEW

**Comment by C0derK1d** [comment]
> Review Status

Yes, I forgot to add a check that is it a barrier or a light block. Sorry about that.
I going to fix it and make a description of this hack.

**Comment by C0derK1d** [comment]
I realized that we don't need a check for light blocks because it indicates barrier blocks only.

File: src/main/java/net/wurstclient/mixin/ClientWorldMixin.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	@Shadow
+	@Final
+	private MinecraftClient client;
+	
+	/**
+	 * This is the part that makes BarrierESP work.
+	 */
+	@Inject(at = @At("HEAD"),
+		method = "getBlockParticle()Lnet/minecraft/block/Block;",
+		cancellable = true)
```
[the remark this answers is not in the document]
Comment (Alexander01998) [reply src/main/java/net/wurstclient/mixin/ClientWorldMixin.java:37]: That's an AI hallucination. The current signature is correct and the proposed one makes no sense. @nobotreply

File: src/main/java/net/wurstclient/mixin/ClientWorldMixin.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	private void getBlockParticle(CallbackInfoReturnable<Block> cir)
+	{
+		if(!WurstClient.INSTANCE.getHax().barrierEspHack.isEnabled())
+			return;
```
[the remark this answers is not in the document]
Comment (Alexander01998) [reply src/main/java/net/wurstclient/mixin/ClientWorldMixin.java:41]: Final field. Can't be null. @nobotreply

File: src/main/java/net/wurstclient/mixin/ClientWorldMixin.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+			
+		// Pause BarrierESP when holding a light in Creative Mode, since it
+		// would otherwise prevent the player from seeing light blocks.
+		if(client.interactionManager.getCurrentGameMode() == GameMode.CREATIVE
+			&& client.player.getMainHandStack().getItem() == Items.LIGHT)
+			return;
+		
+		cir.setReturnValue(Blocks.BARRIER);
```
[the remark this answers is not in the document]
Comment (Alexander01998) [reply src/main/java/net/wurstclient/mixin/ClientWorldMixin.java:49]: Nope. `CallbackInfoReturnable` doesn't throw if it's cancelled multiple times. Besides, the only place where this `CallbackInfoReturnable` instance can be cancelled is in the `ClientWorldMixin.getBlockParticle()` method.

Actually, now that you mention it, this method should be called `onGetBlockParticle()`.

@nobotreply

File: src/main/java/net/wurstclient/mixin/ClientWorldMixin.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	private void getBlockParticle(CallbackInfoReturnable<Block> cir)
+	{
+		if(!WurstClient.INSTANCE.getHax().barrierEspHack.isEnabled())
+			return;
+			
+		// Pause BarrierESP when holding a light in Creative Mode, since it
+		// would otherwise prevent the player from seeing light blocks.
+		if(client.interactionManager.getCurrentGameMode() == GameMode.CREATIVE
+			&& client.player.getMainHandStack().getItem() == Items.LIGHT)
+			return;
```
[the remark this answers is not in the document]
Comment (Alexander01998) [reply src/main/java/net/wurstclient/mixin/ClientWorldMixin.java:47]: True, but `ClientWorld.getBlockParticle()` accesses the same fields without null checks. Therefore it's safe to assume that they won't be null. @nobotreply

**Review by Alexander01998 — APPROVED** [review]

**Pull request merged**